Tutorial Writeup - Merge Sort #91 issue solved

Sreetama2001 opened the pull request:
Tutorial Writeup - Merge Sort #91 issue solved

Added full documentation of Merge Sort with detailed illustration and explanation.

wingkwong reviewed (changes requested):
The content is copied from https://www.simplilearn.com/tutorials/data-structure-tutorial/merge-sort-algorithm

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -47,47 +47,68 @@ Then we can clearly see that merge sort is performed like this way.
 
-  - Compare the element for each list and then combine them into another list in a sorted manner.
-  - Remember each half is already sorted so we need to just **combine/merge** then to make a bigger sorted array
+  - Compare the element for each subarray /list  and then combine them into another bigger list/array in a sorted manner (Here we need to have an idea of 2 pointer $i.e$ how to merge and increment ).
```
what's the point of applying LaTex on `i.e`?

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -47,47 +47,68 @@ Then we can clearly see that merge sort is performed like this way.
+  - Remember each half is already sorted so we need to just sort between 2 halves to **combine/merge** then to make a bigger sorted array
   
   
   ![image](https://miro.medium.com/max/1400/1*_QpN-m-FTRabFLRvtcc04A.gif)
```
If this is not made by you, please remove it.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -47,47 +47,68 @@ Then we can clearly see that merge sort is performed like this way.
-  ```
-  MergeSort(arr[], l,  r)
-If r > l
+[*Leetcode Problem 912 Sort an Array*](https://leetcode.com/problems/sort-an-array/)
````
Starting this line, the format is really bad.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -47,47 +47,68 @@ Then we can clearly see that merge sort is performed like this way.
 ```
-    
+  
 ### C++ Implementation 
````
For implementation with different languages, use something like this

````
<Tabs>
<TabItem value="cpp" label="C++">

```cpp
// YOUR CODE GOES HERE
```

</TabItem>

<TabItem value="py" label="Python">

```py
# YOUR CODE GOES HERE
```

</TabItem>
</Tabs>
````

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -211,76 +239,267 @@ Call `mergeSort` to use this template to sort by divide and conquer as explained
 
 #### Bottom Up Approach 
 
  ![image](https://www.simplilearn.com/ice9/free_resources_article_thumb/mergesort/merge_sort-bottom-up-img1.png)
```
don't copy images from others

Sreetama2001 replied:
Should I remove all the images? Many of them are taken from Medium articles.

wingkwong replied:
Yes. Don't copy other's work.

wingkwong reviewed (commented):
1. Run the app locally and verify before requesting review. I can't even start the app with your broken changes.
2. The format is pretty bad, in terms of the content structure and the coding part.
3. Don't use other images from other sites. Give credits if you have to use them.

Sreetama2001 commented:
> 1. Run the app locally and verify before requesting review. I can't even start the app with your broken changes.
> 2. The format is pretty bad, in terms of the content structure and the coding part.
> 3. Don't use other images from other sites. Give credits if you have to use them.

Why this is breaking I can't figure it out !! Though I have now all necessary changes

Sreetama2001 commented:
@wingkwong  almost formatted it to my best! 
Please have a look.

wingkwong reviewed (changes requested):
1. When you change something based on my comment, click `resolve conversation`.
![image](https://user-images.githubusercontent.com/35857179/193435967-4136bf21-84ca-489d-873e-81435a5cb821.png)

2. Run the app locally and verify first. See [Quick Start](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#quick-start)

 > Why this is breaking I can't figure it out !! Though I have now all necessary changes
![image](https://user-images.githubusercontent.com/35857179/193435938-34573d0f-2a3a-4148-9e7b-fcd445a1a41d.png)

I won't review if I can't even start the app.

Sreetama2001 commented:
@ app is running well good ! please check for the other formatting changes

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+```cpp
+// call mergeSort(nums, 0, nums.size() - 1) to sort 
+
+void merge(vector<int>& nums, int l, int m, int r){
````
Format the intentation in IDE and add it here.

vigneshshiv replied:
Intentation has to be corrected. Use IDE instead of formatting in markdown editor. 

![image](https://user-images.githubusercontent.com/10242660/193452850-4310c93a-5275-413c-be62-deebdfeef48f.png)

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+    mergeSort(nums, l, mid);
+    mergeSort(nums, mid + 1, r);
+    merge(nums, l, r);
+}
```
Leave a line break after the method end

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+
+### Algorithm 
+
+```
````
Format needed, more than one space and tab space need to remove
Eg; it  starts  should it starts

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+<TabItem value="java" label="Java">
+
+```java
+class Solution {
````
Leave a line break after method end. 
Why mergeSort2 and merge2 ?

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+</TabItem>
+</Tabs>
+
+**Time_Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
```
Why _ ? 
Use Time complexity: and Space complexity: 
Correct : as well.

vigneshshiv replied:
Why undercore ??? remove it

![image](https://user-images.githubusercontent.com/10242660/193453079-42a68412-5fce-4920-8c6a-97d379b59779.png)

vigneshshiv replied:
Still, you havn't removed _ in Time and space complexity. 

Time complexity
Space complexity

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+
+**Space_Complexity** :  $O(n)$ 
+
+**Stablitiy**       :  Stable
```
Stability spell

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+
+**Stablitiy**       :  Stable
+
+**Is In Place**      :  Not-In-Place
```
Not-in-place

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+</Tabs>
+
+
+## Bottom Up Approach /Iterative technique
```
Approach / Iterative

vigneshshiv replied:
Still not ok, after / give a space like this.. Approach / Iterative

![image](https://user-images.githubusercontent.com/10242660/193452927-18f7dcb4-cb19-410f-9e37-9a42ca7800fa.png)

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+**Is In Place**      :  Not-In-Place
+
+**When to use**     :  
+1. When we don't have random access (linked list) , like as we have an array.
```
(linked list),

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+
+**Problem Statement** 
+
+`Given the head of a linked list, return the list after sorting it in ascending order.`
```
Remove the `` quote for problem statement

vigneshshiv replied:
Remove ``` for problem statement

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+Output: [-1,0,3,4,5]
+```
+
+Here we can follow both top down or bottom up merge sort . I have already discussed the algorithms in detailed way for array .
````
full stop before space to be removed

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+```
+
+
+**Leetcode Implementation of Linked List in cpp**
````
What is the purpose of this code block ?

Sreetama2001 replied:
I thought it would describe the implementation of LL shortly 

this block of code is given by leetcode
to understand the structure of LL @vigneshshiv

vigneshshiv replied:
@Sreetama2001 Add it as part of the code above or below, instead of separate block

Sreetama2001 replied:
ohk

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+
+### **Steps to apply merge sort to a Linked list** 
+    
+**Note. Implementation is basically based on Merge Sort on an array as discussed above.** 
```
Note:

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,519 @@ keywords:
+**Note. Implementation is basically based on Merge Sort on an array as discussed above.** 
+
+1. Divide the linked list into two equal parts until when only one element is left
+2. To Divide, we need to  find mid in linked ling using [tortoise and hare](https://dev.to/alisabaj/floyd-s-tortoise-and-hare-algorithm-finding-a-cycle-in-a-linked-list-39af) method 
```
Cititation relavent to this problem here ?

Sreetama2001 replied:
@vigneshshiv  just added here if the user don't know about the tortoise and hare algo

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+```cpp
+class Solution {
+public:
+    void mergeSort(ListNode **head){
````
Format the code in IDE and add it. 
Remove unnecessary comments.

Sreetama2001 replied:
ohk

vigneshshiv reviewed (commented):
Correct the changes

Sreetama2001 commented:
@vigneshshiv  @wingkwong  please have a look 

What I have done 

1. Removed the indentation errors
2. Removed typos 
3. Removed necessary spaces, adjusted the punctuation
4. Improved my grammar
5. Remove a few comments 
6.  checked if any more links are broken and changed the tortoise & hare reference link

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,512 @@ keywords:
+
+    Similarly, we then merge the sorted subarrays like we have done in top-down recursive approach (two-pointer approach) 
+
+4. Continues until we have a sorted array 
```
Where is 3 ?

This part need to change, don't use ```. Remove 237 and 247 which has ```
Why there is an extra space in this line -- It is an iterative app. Don't leave a tab space.

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,512 @@ keywords:
+        mergeSort2(nums);
+        for (int i : nums) res.add(i);
+        return res;
+    }
```
Leave a bline break after method end.. 
after 259, add empty line if method end. Use the same for below as well.

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,512 @@ keywords:
+
+### **Steps to apply merge sort to a Linked list** 
+    
+**Note. Implementation is based on Merge Sort on an array as discussed above.** 
```
Correct **Note:**

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,512 @@ keywords:
+        //call to merge left and right by sorting them
+        *head=merge(left,right);
+        
+    }
```
Empty line after method over. 
It's better for readability.

vigneshshiv reviewed (commented):
Overall styling, grammer is Ok, but format and intentation and few changes required.

Sreetama2001 commented:
@vigneshshiv  

> Why undercore ??? remove it

Here for base 2 that's the syntax

Sreetama2001 commented:
@vigneshshiv almost  the changes are done, sorry for the inconvenience 
Please check !

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,531 @@ keywords:
+
+```cpp
+void stableStlSort(vector<int>& nums){
+    
````
You can remove this empty line here.

Sreetama2001 replied:
empty lines  after braces should be removed in a method ? @vigneshshiv

vigneshshiv replied:
It's not after braces like that, when you have few lines after method open brace, you can leave a empty line. 
But for only one line, it's not required to do it. 
Just remove line 87 empty line - that's enough

Sreetama2001 replied:
ohk done formatting! It's my first time wring a blog that's why I have so many errors! Sorry for the inconveniences

vigneshshiv commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,531 @@ keywords:
+</Tabs>
+
+
+## Bottom Up Approach/ Iterative technique
```
Apprantely, you have placed it wrongly. Use the below one

Bottom Up Approach / Iterative technique

Sreetama2001 replied:
should  ##  be removed?

vigneshshiv replied:
Noo, I was telling about /, It shouldn't be **Approach/ Iterative**, you see 1 space before /. 
Use below that's enough

## Bottom Up Approach / Iterative technique

vigneshshiv reviewed (commented):
3 changes in this review, Please look carefully once.

Sreetama2001 commented:
@vigneshshiv  Please check it

vigneshshiv reviewed (commented):
LGTM! @wingkwong request you to do final review.

Sreetama2001 commented:
@wingkwong  can you please review  my PR

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,521 @@ keywords:
+Merging of n elements takes n time and since each time the array is cut into half it takes   $\log_{2} n$   time to reach the top so **total time complexity is**    $O(n\log_{2} n)$.
+
+
+![image](https://cdn.kastatic.org/ka-perseus-images/5fcbebf66560d8fc490de2a0d8a0e5b1d65c5c54.png)
```
if you don't make this image, then don't use it. or give credit

Sreetama2001 replied:
I guess the guy who has used this image also picked it up from somewhere! 
Still, I will give the credits but the he doesn't deserve it I guess

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,521 @@ keywords:
+
+```
+
+## **Approach for top-down Merge sort of Linked list**
````
if you use header, it's already bold. there is no need to use ``**`` again.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,521 @@ keywords:
+*2. Sort each sublist and merge each sorted sublist in a new array.*
+
+
+### **Steps to apply merge sort to a Linked list** 
```
same as above.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,521 @@ keywords:
+        ListNode* slow = curr; // make a slow pointer
+        ListNode* fast = curr -> next; // make a fast pointer
+        
+    // then we move our fast up to it not become null
```
indentation again .. i expect this kind of stuff can be handled well before asking for a review

wingkwong replied:
bad indentation. don't resolve if you haven't fixed

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,521 @@ keywords:
+*Here `insertion sort`, `quick sort`, `selection sort` and all other sorts in $O(n^2)$ will give **TLE** i.e Time Limit Exceeded*
+
+## STL Approach  
+Note: Not preferred here in this problem, but still you will get an AC in Leetcode
```
why is it not preferred? elaborate more for beginners.

Sreetama2001 replied:
It is already mentioned in the question that do not not use stl

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,521 @@ keywords:
+
+Then we can see that merge sort is performed in this way. 
+
+![image](https://miro.medium.com/max/600/1*opwN0BhtH4zvPF697fPlow.gif)
```
if you didn't make this gif, remove it or give credit.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,521 @@ keywords:
+### Algorithm 
+    
+```
+MergeSort(array[], left,  right)
+If right > left
+
+Find the middle point to divide the array into two halves: 
+middle = left + (right – left)/2
+Call mergeSort again for first half for further dividing:   
+Call mergeSort(array, left, middle)
+Call mergeSort again  for second half for further dividing:
+Call mergeSort(array, middle+ 1, right)
+Merge the two halves sorted in steps 2 and 3:
+Call merge(arrary, left, middle, right)
````
prefer using list to display

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,521 @@ keywords:
+### Algorithm 
+
+```
+1. So it  starts with an  element in the array 
````
why there is some extra space ..

wingkwong reviewed (changes requested):
1. put comment above the line. don't do it inline. the idea is not let users to scroll to see the full comment and keep it consistent with other pages.

e.g. 

```
ListNode* res = NULL; // set answer to null
```

to

```
// set answer to null
ListNode* res = NULL;
```
2. format your code ...

3. those comments made by @vigneshshiv are not resolved yet. and there is quite a lot of formatting issues. I highly suggest you to take a look and fix them first before asking for a review. You should be responsible for reviewing first.

Sreetama2001 commented:
> 1. put comment above the line. don't do it inline. the idea is not let users to scroll to see the full comment and keep it consistent with other pages.
> 
> e.g.
> 
> ```
> ListNode* res = NULL; // set answer to null
> ```
> 
> to
> 
> ```
> // set answer to null
> ListNode* res = NULL;
> ```
> 
> 2. format your code ...
> 3. those comments made by @vigneshshiv are not resolved yet. and there is quite a lot of formatting issues. I highly suggest you to take a look and fix them first before asking for a review. You should be responsible for reviewing first.
> 4. resolve the convo if you fix that related issue. Don't leave it open.

I have already followed all the points you asked me to do!

Sreetama2001 commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,519 @@ keywords:
+```
+**Constraints as given in the problem:**
+
+$1 <= nums.length <= 5 * 10^4
````
cant use if it is added a code and as per the template in leetcode cant add bullets

Sreetama2001 replied:
Added the constraints with ```

Sreetama2001 reviewed (commented):
I have resolved all the changes you @vigneshshiv have asked for.

Sreetama2001 commented:
@wingkwong  kindly go through it.

Sreetama2001 reviewed (commented):
All the changes are done and resolved @wingkwong

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+</TabItem>
+</Tabs>
+
+**Time Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
```
because?

Sreetama2001 replied:
Time complexity is explained later @wingkwong

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+
+**Stablity**       :  Stable
+
+**Is In Place**      :  Not in place
```
same as above. Why suddenly mention this while you haven't introduced above.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+
+**Is In Place**      :  Not in place
+
+**When  not to use**: When the array is too large
```
why? state the reason here. and better state the suggestion of what to use.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+**When  not to use**: When the array is too large
+
+
+Merging of n elements takes n time and since each time the array is cut into half it takes $\log_{2}n$ time to reach the top so **total time complexity is** $O(n\log_{2} n)$.
```
`n` -> $n$

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,504 @@ keywords:
+3->8->12->null
+1->10->23->null
+
+After merging we get :
+
+1->3->8->10->12->23->null.
+
+------ Hence Sorted---------------------------
+
+1. The two lists to be merged must be ordered respectively.
+2. We can only start to merge two lists that only have one element.
+3. Then we get an ordered list that has two elements
+4. Do this again (that is "recursion").
+
+ Bottom-Up Approach 
+
+    -1-> 5->  3-> 4->   0-> null
+      \  /       \ /         /
+   -1->5->null 0->3->4->null 
+      \        /         
+    -1-> 0->3-> 4-> 5->null   (sorted) 
+
+----- Finally Sorted ---------------------------------
+
+ Top Down Approach
+
+   -1-> 5->  3-> 4-> 0-> null
+      /          \ 
+-1-> 5->null      3-> 4->0->null
+ /       \           /      \  
+-1->null 5->null 3->null   4->0->null
+  \   /        \          /     \
+ -1-> 5->null  3->null   4->null 0->null
+   \             \        \   /
+    \           3->null     0-> 4->null
+     \              \      /  
+      \           0->3->4->null
+       \          /
+    -1->0->3->4->5->null    (sorted)
+```
````
the presentation of this part is not that nice. Especially the linked list part. Suggest to make some diagrams.

Sreetama2001 replied:
@wingkwong I am resolving this part by removing it! Since drawing diagrams needs to be done by Mermaid, I guess GitHub supports it, but it's too time-consuming to draw a tree with a linked list.

wingkwong replied:
What is Mermaid? 

You can use lucidchart or draw.io to draw the diagram and upload as images and paste the link there.

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+
+**Note: Implementation is based on Merge Sort on an array as discussed above.** 
+1. Divide the linked list into two equal parts until when only one element is left.
+2. To Divide, we need to  find mid in the linked list using **slow and fast pointers** method.
```
unnecessary space before `find mid`

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+
+        // if the linked list is null 
+        // or the size is one return back the same
+        if(curr ==NULL || curr ->next==NULL) return; 
```
poor format. use 

```
if(curr == NULL || curr-> next == NULL)
```

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+        if(curr ==NULL || curr ->next==NULL) return; 
+
+        //call to find the middle node between left and right
+        findMiddle(curr, &left,&right);
```
add space. i.e.

```
findMiddle(curr, &left, &right);
```

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+        // or the size is one return back the same
+        if(curr ==NULL || curr ->next==NULL) return; 
+
+        //call to find the middle node between left and right
```
space .. keep it consistent with other comment styles

```
// call to find the middle node between left and right
```

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+        // call merge_sort again to divide right half
+        mergeSort(&right);
+        
+        //call to merge left and right by sorting them
```
space. same as above

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+        if(right == NULL) return left;
+    
+        // if value of the left <= value of right
+        //then res = left
```
space

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+    
+        // if value of the left <= value of right
+        //then res = left
+        if(left -> val <= right -> val) 
```
don't add space for arrow operator. please change all.

`left->val` and `right->val`

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+    // index for temporary array
+    int k = 0; 
+    while(i <= m && j <= r){
+        //increment the left pointer 
+        //if the right pointer element is bigger 
+        //Since we are sorting in ascending order,left(smaller element) goes first
```
space after //

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,504 @@ keywords:
+4. Merge the two halves sorted :
+    Call merge(arrary, left, middle, right)
+```
+<Tabs>
````
1. i expect to see same detailed comments in 3 different langs. 
2. adopt in LC solution format. Don't just show 2 functions. (i.e. i can copy and paste to run in LC directly)

Sreetama2001 replied:
@wingkwong  this is the generalized algorithm for merge sort, So how could I add comments in different languages and introduce LC format?

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
```diff
@@ -11,4 +11,504 @@ keywords:
+        if(nums[i] <= nums[j]) tmp[k++] = nums[i++];   
+        else tmp[k++] = nums[j++];
+    }
+    while(i <= m) tmp[k++] = nums[i++];
+    while(j <= r) tmp[k++] = nums[j++]; 
+    for(i = 0; i < k; i++) nums[l + i] = tmp[i];
```
add comments for these 3 lines

wingkwong commented on tutorials/basic-topics/sorting/merge-sort.md:
````diff
@@ -11,4 +11,504 @@ keywords:
+
+```java
+class Solution {
+    public List<Integer> sortArray(int[] nums) {
````
in LC problem, it is

```
public int[] sortArray(int[] nums) {
```

please rewrite this solution.

Sreetama2001 commented:
https://github.com/wingkwong/leetcode-the-hard-way/pull/102/files/53dcd3e6df6ca4524c61ba3a4f5d8d3576b5cebe#r986395310

@wingkwong  For the Algorithm part, I left it unresolved.
Also I have removed the diagram part since it needed to be done with Mermaid but since it is a linked list with tree diagram it requires too much time to be invested (reading the docs and implementing it several times to make it the best fit)

Sreetama2001 reviewed (commented):
Resolved

Sreetama2001 commented:
> What is Mermaid?

[mermaid](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) - It is in-built in github for diagrams

> You can use lucidchart or [draw.io](http://draw.io/) to draw the diagram and upload it as images and paste the link there.
Do you need those diagrams? I mean it will be a little time-consuming for me.
So are those diagrams worth it?
I have made the other changes, So when you will get time please take note of those 

- Detailed comments have been added
- yes  `#` is been used in python
- added class `Solution` as Leetcode
- added the method names as per Leetcode
- If there is a space for more improvement I will do it 
- For the diagrams and stuff I need more time, so I would be grateful if you allow a separate PR for that (if that's not for hacktober fest it would be also fine)

wingkwong commented:
Check discord for the comments.

wingkwong reviewed (approved):
Looks much better than before, even though there's some improvement. I'll change it myself.
